File: output/copilot/javascript/hard/q37-longest-valid-parentheses.js
/** https://leetcode.com/problems/longest-valid-parentheses */
// Problem: Longest Valid Parentheses

//Problem Description:
//Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses substring.

// Constraints:
// 0 <= s.length <= 3 * 10^4
// s[i] is '(', or ')'.


//Code Structure:
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
    let maxLength = 0;
    const stack = [-1]; // Initialize stack with -1 to handle base case

    for (let i = 0; i < s.length; i++) {
        if (s[i] === '(') {
            stack.push(i); // Push the index of '(' onto the stack
        } else {
            stack.pop(); // Pop the last index from the stack
            if (stack.length === 0) {
                stack.push(i); // If stack is empty, push current index as a new base
            } else {
                maxLength = Math.max(maxLength, i - stack[stack.length - 1]); // Calculate valid length
            }
        }
    }

    return maxLength; // Return the maximum length of valid parentheses substring   
};
    

File: output/copilot/java/hard/q37-longest-valid-parentheses.java
/** https://leetcode.com/problems/longest-valid-parentheses */
// Problem: Longest Valid Parentheses

//Problem Description:
//Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses substring.

// Constraints:
// 0 <= s.length <= 3 * 10^4
// s[i] is '(', or ')'.

//Code Structure:
class Solution {
    public int longestValidParentheses(String s) {
        int maxLength = 0;
        int lastInvalidIndex = -1; // To track the last index of an invalid parenthesis

        // Stack to keep track of indices of '(' characters
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (c == '(') {
                stack.push(i); // Push the index of '(' onto the stack
            } else {
                if (stack.isEmpty()) {
                    // If we encounter a ')' without a matching '(', mark this as invalid
                    lastInvalidIndex = i;
                } else {
                    // Pop the matching '(' index from the stack
                    stack.pop();
                    if (stack.isEmpty()) {
                        // If the stack is empty, calculate the length from last invalid index
                        maxLength = Math.max(maxLength, i - lastInvalidIndex);
                    } else {
                        // If the stack is not empty, calculate length from the top of the stack
                        maxLength = Math.max(maxLength, i - stack.peek());
                    }
                }
            }
        }

        return maxLength; // Return the maximum length of valid parentheses substring found   
    }
}
    